README.md
# mcanvas

New version 2.0.0 is coming. It contains some new features:

- Support new function:
    - image merge；
    - image compress；
    - image crop;
    - image filter;
- Support usage in node.js；
- .draw() support usage by promise;
- Rewrite in typescript;

 **[Example](http://guoxiaodong.net/)**

 **[Git](https://github.com/xd-tayde/mcanvas)**

## Document

[English](https://github.com/xd-tayde/mcanvas/blob/master/README_EN.md) | [中文版](https://github.com/xd-tayde/mcanvas/blob/master/README_ZH.md)

## Introduction

mcanvas is a image handler plugin that can easily merge, crop, compress, filter the image and export a image of base64 finally. It provides some simple api that based on canvas, in order to make your work more efficiently and conveniently.

## Installation

- You can download the latest version from the [GitHub](https://github.com/xd-tayde/mcanvas/blob/master/dist)
- use a npm CDN [Web](https://unpkg.com/mcanvas/dist/mcanvas.web.js) and use `window.MCanvas`

- Or you can install via npm:

```js
npm install mcanvas --save

import { MCanvas } from 'mcanvas'
```

## Basic Usage

- `MCanvas`

```js
// create the canvas by width and height;
import { MCanvas } from 'mcanvas'

const mc = new MCanvas({
	width,
	height,
	backgroundColor,
});

// prepare background-image
mc.background(image, {
    left: 0,
    top: 0,
    color: '#000000',
    type: 'origin',
})

// prepare the image material, add into queue;
.add('images/nose.png',{
    width:183,
    pos:{
        x:250,
        y:369,
        scale:0.84,
        rotate:1,
    },
})

// add text;
.text('normal<br><s>smallsmall</s>',{
    width:'300px',
    align:'center',
    pos:{
        x:0,
        y:0,
    },
})

// prepare watermark;
.watermark(img ,{
    width:'40%',
    pos:'rightBottom',
})

// draw all material that prepared before, and get the base64-image
.draw(b64 =>{
    console.log(b64);
});
```
- `MImage`

```js
// image handler
    // inclue crop / compress / common filter
import { MImage } from 'mcanvas'

const mi = new MImage('http://mtapplet.meitudata.com/596c72073971d86b5128.jpg')

mi.filter('blur')
// crop to area by 300 * 300 and center in origin image
.crop({
    x: 'center',
    y: 'center',
    width: 300,
    height: 300,
    radius: 10,
})
// compress into a image that width is 200px and quality is 0.9
.compress({
    width: 200,
    quality: .9,
})
// get the base64-image
.draw(b64 => {
    console.log(b64)
})
```

## License

[MIT](https://opensource.org/licenses/MIT)


CHANGELOG.md
## 更新：
- 2.0.8
    - MCanvas.text 新增 wordBreak 属性，用于设置自动换行点；
    - 更新文档；
    - 修复声明文件错误；

- 2.0.0
    - 重整项目架构，更好的模块化代码，引入 Typescript， Promise;
    - 重写 MCrop, 优化实现方式，新增 radius；
    - 原 MCanvas 调整为 MComposer， 
        - 优化 `mc.rect`，新增 radius；
        - 优化 `mc.add.crop`， 新增 radius;
        - `mc.draw` 方法新增 exportConfig： 'base64' | 'canvas' 配置导出格式；
    - 抽离 Queue 类；
    - 新增 MFilter 滤镜器：
        - 高斯模糊；
        - 翻转；
        - 灰白；

- 1.3.9 (2019-5-10)
    - 优化文档；
    - 修复 MCrop 裁剪时图片格式的设置问题；
    - 优化 rollup 打包，导出 MCrop 方法；

- 1.3.8 (2019-5-9)
    - 升级新版 rollup 脚手架 (rollup-typescript-startkit);
    - 新增图片裁剪工具 MCrop, 更新文档;
    - 修复小数点计算问题导致图片变模糊的问题；
    
- 1.3.4 (6.22)
    - 修改文字绘制机制，新增文字旋转参数，`option.pos.rotate`；

- 1.3.2 (4.28)
    - `add`函数`width`参数新增`origin`, 表示按原图尺寸绘制；
    - 文字绘制新增容错: 当 `options.width < fontSize` 时，则按`fontSize`进行绘制，避免换空行的问题；
    - 修复初始化不传参数报错的问题；

- 1.3.1 (4.28)
    - 修复`text`调用多次会相互影响的问题；

- 1.3.0 (2.22)
    - 修改新建实例时的参数格式，较为统一且便于拓展；
    - 优化储存管理分类，实例上新增 `fn` 及 `data` 属性;
    - 新增Api `mc.rect` / `mc.circle`, 用于绘制矩形和圆形；
    - 纠正`mc.text`方法定位不准确的问题，并新增描边文字,渐变文字，文字阴影的功能;
    - 新增错误回调，当图片加载失败是会触发由 `draw` 中传入的 `error`回调;
    - 优化文档及代码；

- 1.2.9 (12.20)
    - 新增Api `mc.clear`: 清空画布；

- 1.2.8 (10.27)
    - 修复`background` API接口的`left/top`参数纵向居中问题，另外新增支持`left:'center'`,统一API;

- 1.2.6 (9.27)
    - `background` API接口的`left/top`参数新增`0%`、`50%` 和 `100%`;

- 1.2.5 (9.27)
    - 修复文档错误；

- 1.2.3 (9.26)
    - 增加素材`canvas`大小的限制，ios8为2096，其余限制为4096；

- 1.2.2 (9.25)
    - 增加参数的向下取整；

- 1.2.1 (9.7)
    - 修改 `add` 函数中的 `crop` 裁剪功能BUG；

- 1.2.0 (8.28)
    - 修改`mc.draw()`的传参形式，增加可配置的导出图片格式及图片质量，兼容旧版`api`;
    - 增加一些可能引发错误的错误提醒，便于业务方定位问题；

- 1.1.8 (8.13)
    - 更新文档；

- 1.1.7 (8.12)
    - 增加`mc.background(bg)`接口初次调用时的错误提醒，修改异常提示的内容；

- 1.1.6 (8.11)
    - 修复竖直素材下,为了规避裁剪的画布放大倍数错误的问题；

- 1.1.5 (8.10)
    - 增加 `add/background/watermark` 传入的`image`可以为 `canvas`；

- 1.1.4 (8.9)
    - 增加英文文档；
    - 增加初始化时可配置默认背景色；

- 1.1.3 (8.3)
    - 修改一处警告语法错误；
    - 增加画素材的小画布放大机制，用于解决因旋转导致的裁剪；
    - 增加小画布绘制后清空机制，避免因素材太大导致占用内存过高；

- 1.0.9 (7.27)
    - 获取图片宽度由 `img.width` 修改为 `img.naturalWidth`,以避免直接传入`DOM`节点时获取宽高被`css`影响；
    - 增加背景图片错误提醒；

- 1.0.8 (7.27)
	- 调整构造函数结构，使用 `prototype` 原型链的形，以兼容不带 `new` 的调用方式；
	- 调整完善文档；
    - 删除错误的包依赖，调整包名；

- 1.0.2 (7.26)
	- 修正完善文档
	- 修改 bigStyle --> largeStyle;

example/web.ts
import { MCanvas, MImage } from '../src/index'
import ear from './images/ear.png'
import watermark from './images/watermark.jpg'
import imgTest from './images/1.jpg'
import './main.scss'

// (async () => {
//     const mc = new MImage('http://mtapplet.meitudata.com/596c72073971d86b5128.jpg')
//     const b64 = await mc.crop({
//         width: 300,
//         height: 300,
//         x: 0,
//         y: 0,
//     }).filter('blur').compress({
//         width: 100,
//         height: 100,
//     }).draw()
//     $('.js-result').attr('src', b64)
// })()

let $sure = $('.js-sure')
let $params = $('.js-params')
let $dialog = $('.js-dialog')
let $result = $('.js-result')
let $cancel = $('.js-cancel')
let $clear = $('.js-clear')

let data = {
    addImageOps : {
        image: ear,
        options: {
            width: 482,
            // crop: {
            //     x: 192,
            //     y: 84,
            //     width: 365,
            //     height: 365,
            //     radius: '50%',
            // },
            pos: {
                x: 150,
                y: 50,
                scale: 1,
                rotate: 0,
            },
        },
    },
    addWmOps : {
        image: watermark,
        options: {
            width: '40%',
            pos: 'rightBottom',
        },
    },
    addRectOps: {
        x: 0,
        y: 'bottom:0',
        width: '100%',
        height: 300,
        radius: 30,
        strokeWidth : 5,
        strokeColor: '#996699',
        fillColor: 'rgba(0,0,0,.5)',
    },
    addCircleOps: {
        x: 'center',
        y: 'center',
        r: 100,
        strokeWidth : 5,
        strokeColor: '#996699',
        fillColor: 'rgba(0,0,0,.5)',
    },
    addTextOps : {
        // text:'<b>A</b>BBBBB<s>MCanvas.js</s>',
        text: '<b>Large/Stroke</b><br>Normal/Gradient<br><s>Small/Shadow</s>',
        options: {
            width: '600',
            align: 'center',
            largeStyle: {
                color: 'red',
                font: '90px Microsoft YaHei,sans-serif',
                type: 'stroke',
                lineWidth: 2,
                lineHeight : 90,
            },
            normalStyle: {
                color: 'blue',
                font: '70px Microsoft YaHei,sans-serif',
                lineHeight : 70,
                // shadow:{
                //     color: 'red',
                //     blur: 4,
                //     offsetX: 2,
                //     offsetY: 2,
                // },
                gradient: {
                    type: 2,  // 1: 横向渐变； 2: 纵向渐变；
                    colorStop: ['red', 'blue'],
                },
            },
            smallStyle: {
                color: 'yellow',
                font: '50px Microsoft YaHei,sans-serif',
                lineHeight : 50,
                shadow: {
                    color: 'red',
                    blur: 10,
                    offsetX: 5,
                    offsetY: 5,
                },
            },
            pos: {
                x: 'center',
                y: 'bottom',
                rotate: 0,
            },
        },
    },
    crop: {
        x: 'center',
        y: 'center',
        width: 300,
        height: 300,
        radius: 150,
    }
}
let mc = new MCanvas({
    width: 1000,
    height: 1500,
    // backgroundColor: 'black',
})
mc.background('http://mtapplet.meitudata.com/596c72073971d86b5128.jpg', {
// mc.background('http://mtapplet.meitudata.com/59e8765b6492c541.jpg',{
    type: 'origin',
    left: '50%',
    top: '50%',
})

const mi = new MImage('http://mtapplet.meitudata.com/596c72073971d86b5128.jpg')

let timer
$('.Button').on('touchstart', function(this: any) {
    $(this).addClass('taped')
    timer = setTimeout(() => {
        $(this).removeClass('taped')
    }, 2000)
})
$('.Button').on('touchend', function(this: any) {
    $(this).removeClass('taped')
    clearTimeout(timer)
})

$cancel.on('click', () => {
    $dialog.hide()
})

$clear.on('click', () => {
    // mc.clear();
    mc.background().clear().draw(b64 => {
        $result.attr('src', b64)
    })
})

$('.js-addImage').on('click', () => {
    showDialog(`image`, data.addImageOps)
})

$('.js-addWm').on('click', () => {
    showDialog(`watermark`, data.addWmOps)
})

$('.js-addText').on('click', () => {
    showDialog(`text`, data.addTextOps)
})

$('.js-addRect').on('click', () => {
    mcDraw(data.addRectOps, 'rect')
})

$('.js-addCircle').on('click', () => {
    mcDraw(data.addCircleOps, 'circle')
})

$sure.on('click', function(this: any) {
    let ops = $(this).data('ops')
    let type = $(this).data('type')

    mcDraw(ops, type)
})

function mcDraw(ops, type) {
    let img
    switch (type) {
        case `image`:
            img = new Image()
            img.crossOrigin = '*'
            img.onload = async () => {
                const b64 = await mc.add(img, ops.options).draw({
                    type: 'jpg',
                    quality: .9,
                })
                $result.attr('src', b64)
                $dialog.hide()
            }
            img.src = ops.image
            break
        case `watermark`:
            mc.watermark(ops.image, ops.options).draw(b64 => {
                $result.attr('src', b64)
                $dialog.hide()
            })
            break
        case `text`:
            mc.text(ops.text, ops.options).draw(b64 => {
                $result.attr('src', b64)
                $dialog.hide()
            })
            break
        case `rect`:
            mc.rect(ops).draw(b64 => {
                $result.attr('src', b64)
                $dialog.hide()
            })
            break
        case `circle`:
            mc.circle(ops).draw(b64 => {
                $result.attr('src', b64)
                $dialog.hide()
            })
            break
        case `crop`:
            mi.crop(ops).draw({
                type: 'png',
                success(b64) {
                    $('.js-cropResult').css('background-image', `url(${b64})`)
                    $dialog.hide()
                }
            })
            break
        default:
    }

}

function showDialog(type, ops) {
    let tab = `&nbsp;&nbsp;&nbsp;&nbsp;`
    let html
    switch (type) {
        case 'image':
            html = `<li>options: {</li>
                    <li>${tab + tab}width:<input data-type='width' class='js-input input' type='text' value='${ops.options.width}'></li>
                    <li>${tab + tab}pos: {</li>
                    <li>${tab + tab + tab + tab}x:<input data-type='x' class='js-input input' type='text' value='${ops.options.pos.x}'></li>
                    <li>${tab + tab + tab + tab}y:<input data-type='y' class='js-input input' type='text' value='${ops.options.pos.y}'></li>
                    <li>${tab + tab + tab + tab}scale:<input data-type='scale' class='js-input input' type='text' value='${ops.options.pos.scale}'></li>
                    <li>${tab + tab + tab + tab}rotate:<input data-type='rotate' class='js-input input' type='text' value='${ops.options.pos.rotate}'></li>
                    <li>${tab + tab}}</li>
                    <li>}</li>`
            break
        case 'watermark':
            html = `<li>options: {</li>
                    <li>${tab + tab}width:<input data-type='width' class='js-input input' type='text' value='${ops.options.width}'></li>
                    <li>${tab + tab}pos:
                        <select data-type='pos' class='js-select select'>
                            <option>rightBottom</option>
                            <option>rightTop</option>
                            <option>leftBottom</option>
                            <option>leftTop</option>
                        </select>
                    <li>}</li>`
            break
        case 'text':
            html = `<li>text: <span style="font-size:12px;">''&lt;b&gt;Large/Stroke&lt;/b&gt;Normal/Gradient&lt;s&gt;Small/Shadow&lt;/s&gt;''</span></li>
                    <li>options:{</li>
                    <li>${tab + tab}width:<input data-type='width' class='js-input input' type='text' value='${ops.options.width}'></li>
                    <li>${tab + tab}align:
                        <select data-type='align' class='js-select select'>
                            <option>left</option>
                            <option>center</option>
                            <option>right</option>
                        </select>
                    </li>
                    <li>${tab + tab}pos:{
                    <li>${tab + tab + tab + tab}x:<input data-type='x' class='js-input input' type='text' value='${ops.options.pos.x}'></li>
                    <li>${tab + tab + tab + tab}y:<input data-type='y' class='js-input input' type='text' value='${ops.options.pos.y}'></li>
                    <li>${tab + tab}}</li>
                    <li>}</li>`
            break
        case 'crop': 
            html = `<li>options: {</li>
                <li>${tab + tab}width: <input data-class="crop" data-type='width' class='js-input input' type='text' value='${ops.width}'></li>
                <li>${tab + tab}height: <input data-class="crop" data-type='height' class='js-input input' type='text' value='${ops.height}'></li>
                <li>${tab + tab}x:<input data-class="crop" data-type='x' class='js-input input' type='text' value='${ops.x}'></li>
                <li>${tab + tab}y:<input data-class="crop" data-type='y' class='js-input input' type='text' value='${ops.y}'></li>
                <li>${tab + tab}radius:<input data-class="crop" data-type='radius' class='js-input input' type='text' value='${ops.radius}'></li>
                <li>}</li>`
            break
        default:
    }
    $params.html(html)
    $sure.data('ops', JSON.stringify(ops)).data('type', type)
    $dialog.show()
}

$(window).on('input', '.js-input', function(this: any) {
    let $this = $(this)
    let v = $this.val()
    let type = $this.data('type')
    let ops = $sure.data('ops')
    const cls = $this.data('class')

    if (cls === 'crop') {
        switch (type) {
            case 'width':
                ops.width = v
                break
            case 'height':
                ops.height = v
                break
            case 'x':
                ops.x = v
                break
            case 'y':
                ops.y = v
                break
            case 'radius':
                ops.radius = v
                break
            default:
                break
        }
    } else {
        switch (type) {
            case 'width':
                ops.options.width = v
                break
            case 'x':
                ops.options.pos.x = v
                break
            case 'y':
                ops.options.pos.y = v
                break
            case 'scale':
                ops.options.pos.scale = v
                break
            case 'rotate':
                ops.options.pos.rotate = v
                break
            case 'align':
                ops.options.align = v
                break
            default:
        }
    }
    
    $sure.data('ops', JSON.stringify(ops))
})

$(window).on('focus', '.js-input', function(this: any) {
    $(this).addClass('focus')
})
$(window).on('blur', '.js-input', function(this: any) {
    $(this).removeClass('focus')
})

$(window).on('change', '.js-select', function(this: any) {
    let ops = $sure.data('ops')
    let type = $(this).data('type')
    ops.options[type] = $(this).val()
    $sure.data('ops', JSON.stringify(ops))
})

$('.js-cropImage').on('click', () => {
    showDialog(`crop`, data.crop)
})

$('.js-filter').on('click', function(this: any) {
    const $btn = $(this)
    const $par = $btn.parent()
    const type = $btn.data('type')
    mi.filter(type).draw(b64 => {
        $par.css('background-image', `url(${b64})`)
    })
})

